### PR TITLE
feat(android): Added Ti.App "close" event support

### PR DESCRIPTION
**Pull request by jquick-axway**
**JIRA:**
https://jira.appcelerator.org/browse/TIMOB-26542

**Summary:**
- Fires "close" event just before JS runtime terminates.
  - We document that this must be the last event. So, this is the perfect place to do it.
  - Had to modify `V8Object` and `V8Function` to allow V8 calls while in the "released" state.
- KrollRuntime was refactored a bit:
  - Removed `RELAUNCH` state. No longer needed since 8.0.0's intent refactoring and now that Services/BroadcastReceivers won't keep JS runtime alive anymore.
  - Removed `CountDownLatch`. No longer needed since JS runtime can only be ran from UI thread.
  - Modified `V8Object` and `V8Function` to allow V8 calls while `KrollRuntime` is in the `RELEASED` state, which happens when an `OnDisposingListener` is being invoked.
- Moved `cancelTimers()` call from `KrollRuntime` to `TiApplication`.
  - Doing this via listener is a better solution. No longer tightly coupled.

**Note:**
I've noticed that our `TiApplication.softRestart()` does not follow the rules and calls the `V8Runtime`'s dispose/init methods directly. This means our `OnDisposingListener`s won't get invoked. The `Ti.App._restart()` handling still works with my PR (I've tested it), but we may want to look into this in the future.

**Close Event Test:**
1. Build and run the below code on Android.
2. Press the "Back" button.
3. Verify the following got logged: `### App 'close' event was fired.`
4. Relaunch the app.
5. Repeat steps 2 and 3.

```javascript
Ti.App.addEventListener("close", function() {
	Ti.API.info("### App 'close' event was fired.");
});
Ti.UI.createWindow().open();
```

**LiveView Test:**
_This test verifies that our runtime restart changes work okay._
1. Build and run [kitchensink-v2](https://github.com/appcelerator/kitchensink-v2) via LiveView on Android.
2. Verify that the app launches okay.
3. Edit a JS file and click save.
4. Verify that the app re-launches okay.

### CODE REVIEW

**Comment by keerthi1032** [comment]
FR passed. Event fired when back button clicked and kitchen sink works fine with live view. build failed in jenkin. Waiting for CR . 

Operating System
  Name                        = Mac OS X
  Version                     = 10.13.6
  Architecture                = 64bit
Node.js
  Node.js Version             = 8.12.0
  npm Version                 = 6.4.1
Titanium CLI
  CLI Version                 = 5.1.1
Studio Version=
CLI =7.0.10 master 10
Titanium SDK
  SDK Version                 = local 8.1.sdk
Device                        =Samsung s5 android 6, Oneplus 5t android 9
Emulator                   =Pixel 3xl android 8

**Comment by jquick-axway** [comment]
@garymathews, I'm going to push this PR out from 8.1.0 to 8.3.0 since it makes high risk changes to our kroll runtime.

**Comment by sgtcoolguy** [comment]
FR passed, but still needs a CR cc @garymathews @ypbnv

File: apidoc/Titanium/App/App.yml [hunk 1]
```diff
@@ -204,8 +204,8 @@ events:
   - name: close
     summary: Fired by the system when the application is about to be terminated.
     description: It is the last event that can be received before the application gets closed.
-    platforms: [iphone, ipad]
-    since: "1.5.0"
+    platforms: [android, iphone, ipad]
+    since: {android: "8.1.0", iphone: "1.5.0", ipad: "1.5.0"}
```
Comment (garymathews) [inline apidoc/Titanium/App/App.yml]: This needs updating to `8.3.0`

**Review by garymathews — APPROVED** [review]
CR: **PASS**

**Pull request merged**